### DISCUSSION TITLE
About the .mml export (spcecifically its potential usage)

### DISCUSSION THREAD

**Discussion by traditionalgithub** [YuriSizov/boscaceoil-blue#45] [Q&A] [answered]
.mml export function exists inBosca Ceoil and I would like to know its potenial usage. What can we do about the exported .mml? Does there exsists some compiler/interpreter for the .mml?

**Comment by YuriSizov** [comment, marked as answer]
Hey there! It's a bit tricky with MML. It's an old general format for encoding musical compositions, which dates to early days of game development, although some more modern projects also support it as a way for players to create their own music for their characters to play.

The format is inherently non-universal. It's up to the particular synthesizer to interpret it, and there are subtle differences between how different synths do so. **So, in short, the MML exported by Bosca is made to be played by the specific synthesizer that it uses internally, SiON.** You should be able to use these files with tools made with the original [SiON](https://github.com/keim/SiON) (a Flash/Adobe AIR library) or the Godot extension [GDSiON](https://github.com/YuriSizov/gdsion). But I'm not aware of any specific user-facing apps that offer you to play SiON MMLs.

So it's niche.

**Reply by traditionalgithub** [reply]
Thanks your answer! SiON sounds like a powerful and interesting library. 
I am also surprised to see the .ceol file being kept as pure text rather than binary data, that's really amazing!

**Reply by YuriSizov** [reply]
The .ceol file format is historically this way, so I preserved it as is in this version of Bosca. But I will have to change it in the future to add support for new features and address some of the issues. I was thinking about making it binary then, since you can keep things really compact and structured this way. Either way, .ceol is just numbers and perhaps some text in the future, so it should be perfectly hackable either way. We don't need to store samples or anything like that in it.

**Reply by traditionalgithub** [reply]
I think one of the advantage of pure text is that it can be shared publically on web without using web storages, especially when some wellknown web storage providers (google drive, drobox, mega......) are blocked from access in some regions. I often come across some nice tunes/arranges in ytube, only to find the provided file links invalid. 

Would you also consider a ".mml import" function in future versions along with the binarization of .ceol?

**Reply by YuriSizov** [reply]
I'm not sure that the file format helps either way with the sharing problem. Generally, if the extension is unrecognized, it is not allowed at all, or everything is allowed regardless. And I doubt .ceol is widely recognized by third-party services.

That said, there was a request to make the file format more friendly for version control systems (like git), and binary files are inherently not friendly to VCS. So there is a good chance I will keep it a text format, just structured very differently.

> Would you also consider a ".mml import" function

Importing of external formats is difficult, because they are more flexible than Bosca in some aspects. So not every MML can be reasonably importable into Bosca, so it's hard to consider it a worthwhile feature (I'm only supporting MIDI imports because the original Bosca did, but the usefulness of it is very dubious). Note composition itself is specifically problematic when it comes to MML.

That said, I do plan to do something related to MMLs. People have been asking for ways to define custom instruments, and I think the MML format would be perfect for that, because SiON MMLs can contain such definitions. In fact, I made [Glasan FX](https://github.com/YuriSizov/glasan-fx) with that in mind and it can export whatever you have configured as an MML voice definition which I want to support for import in a future version.

May I ask what your use case would be here? Why import MMLs?

**Reply by traditionalgithub** [reply]
Sorry about my previous comment sounds misleading (and also demanding), i was talking about if Bosca would be able to import the .mml produced by Bosca itself with vely limited external modifications (like adding extra comment lines) or with no modifications at all. 

It is more difficult for one in the Chinese Mainland to access cloud storage provider like google or dropbox, and most local providers there are not very nice to use. So i once did a video skit made with a little ppmck snippet on bilibili, and post the ppmck snippet as an article also on bilibili, and the platform seems to be ok with it .
>That said, there was a request to make the file format more friendly for version control systems (like git), and binary files are inherently not friendly to VCS. So there is a good chance I will keep it a text format, just structured very differently.

That's good news!

I have seen video games music made with Bosca, and porting it to Godot makes it very Os (for example, Linux) friendly. This tool is very promising!
